### ISSUE TITLE
Why the loss doesn't decrease in my CNN model?

### ISSUE THREAD

**Issue by lihongxun945** [tensorflow/tensorflow#37537]
Hi All,
I am new to tensorflow, and I find the losses doesn't decrease when train a CNN model on Cifar100 dataset.

Here is the Model code:

```python
network = Sequential([
    layers.Conv2D(64, kernel_size=[3, 3], padding="same", activation=tf.nn.relu), # 32x32
    layers.Conv2D(64, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 16x16

    layers.Conv2D(128, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.Conv2D(128, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 8x8

    layers.Conv2D(256, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.Conv2D(256, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 4x4

    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 2x2

    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 1x1

    layers.Flatten(),

    layers.Dense(256, activation=tf.nn.relu),
    layers.Dense(128, activation=tf.nn.relu),
    layers.Dense(100, activation=None),
])
```

and this is the link for the entire code: https://colab.research.google.com/drive/1kIT_18Z-ymEgzqgG4nh4sZdBkA1bnJdO

and, It will be OK if I use only one Conv2D layer with maxpooling, like this:

```python
network = Sequential([
    layers.Conv2D(64, kernel_size=[3, 3], padding="same", activation=tf.nn.relu), # 32x32
    layers.MaxPooling2D((2, 2)), # 16x16

    layers.Conv2D(128, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 8x8

    layers.Conv2D(256, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 4x4

    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 2x2

    layers.Conv2D(512, kernel_size=[3, 3], padding="same", activation=tf.nn.relu),
    layers.MaxPooling2D((2, 2)), # 1x1

   #xxxxx
])
```

thanks~

**Comment by Eshan-Agarwal** [comment]
@lihongxun945  You are not defining your last layer properly like you set `activation = None` replace it with `activation = "softmax"`

**Comment by lihongxun945** [comment]
@Eshan-Agarwal  Define the last layer property as activation='softmax' will not work, I have tried.

**Comment by Eshan-Agarwal** [comment]
@lihongxun945  try some things like change your optimizer to sgd with momentum or make a big network as cifar-100 is very huge data. Please have a look I train DenseNets on CIFAR-10 dataset
[here](https://github.com/Eshan-Agarwal/DenseNet-on-CIFAR-10-/blob/master/CNN%20on%20CIFAR.ipynb) also keep activation = softmax in last layer

**Comment by Ankuraxz** [comment]
@lihongxun945  You can try dropout layer, you need not change the Adam optimizer, rather try using a bigger network from scratch or use Transfer learning { will share an example real soon}. Finally in last layer Dense(<no of classes>, softmax) is perfect. The decision of the combination of layers is a topic of great research. The accuracy of your model touched 58.8 %, which isn't really impressive, try making a deeper network.

**Comment by Eshan-Agarwal** [comment]
Yes forgot to mention dropouts really worked much better.

**Comment by rmothukuru** [comment]
@lihongxun945,
Please refer this [Stack Overflow Answer](https://stackoverflow.com/questions/57061266/how-to-improve-my-cnn-high-and-constant-validation-error/59172613#59172613) which elaborates on how to decrease the loss. Thanks!

**Comment by Eshan-Agarwal** [comment]
@rmothukuru yes use of early stopping helps alot which stop model to learn too much ie, prevent from overfiting. I also get better result on CIFAR -10 data set using DenseNets with early stopping, got arround 90% accuracy on test data. In above mentioned link

**Comment by lihongxun945** [comment]
@rmothukuru  Thanks~
For your two points:
1. The input shape is defined after the model definition, you can see `network.build(input_shape=[None, 32, 32, 3])` in the entire code here https://colab.research.google.com/drive/1kIT_18Z-ymEgzqgG4nh4sZdBkA1bnJdO
2. I have tried add a 'softmax' layer but it donesn't make any change.

**Comment by lihongxun945** [comment]
Thanks  very much @Ankuraxz and @Eshan-Agarwal , I am a beginner of TF, and don't know how to use dropout right now. But I will try it when complete my online course.

**Comment by Eshan-Agarwal** [comment]
@lihongxun945   Its very simple just use as this `layers.Dropout(rate)` , rate means how many neurons you want to be connected its in between 0 to 1, 1 means all neurons is connected it is same as with no dropouts. You can read about them [here](https://keras.io/layers/core/), and define your dropout layes just as you defining pooling layer.

**Comment by rmothukuru** [comment]
@lihongxun945,
Please refer this [Stack Overflow Answer](https://stackoverflow.com/questions/57061266/how-to-improve-my-cnn-high-and-constant-validation-error/59172613#59172613) which elaborates on how to decrease the loss and improve your accuracy, and also shows how to use the functions like Dropout, Early Stopping, Regularization, Image Data Generator, etc..

**Comment by Ankuraxz** [comment]
Just add <layers.Dropout(0.5)>    0.5 is working great on your project

**Comment by Ankuraxz** [comment]
@lihongxun945  You can try dropout layer, you need not change the Adam optimizer, rather try using a bigger network from scratch or use Transfer learning { will share an example real soon}. Finally in last layer Dense(<no of classes>, softmax) is perfect. The decision of the combination of layers is a topic of great research. The accuracy of your model touched 58.8 %, which isn't really impressive, try making a deeper network.

**Comment by rmothukuru** [comment]
@lihongxun945,
Can you please confirm if your issue is resolved. Thanks!

**Comment by Ankuraxz** [comment]
Yes please let us know, If Transfer Learning on Cifar 100 was of any use

**Comment by lihongxun945** [comment]
@rmothukuru  Already reolved, after I changed the model structure。 Thanks you guys~
